### ISSUE TITLE
Add useCollectionService to handle collection service tags

### ISSUE THREAD

**Issue by YuuwaExpanderBleed** [LastTalon/matter-hooks#16]
## Problem
As someone who uses collection service tags a lot to mark instances to have certain component, I thought it may be useful to add useCollectionService hook to handle them.

The advantage of this is that unlike useEvent, this will also handle all the instances with tag already added, meaning you won't need to run same operation twice separately for existing instances and newly tagged instances. 

## Proposed solution
The implementation of this hook is quite identical to useStream since it both handles addition and removal of instance, except on different event. It also has to add instances with tag already added on initialization.

## Alternatives
One alternative is to have a similar implementation to what is provided in the matter docs. 
https://eryn.io/matter/docs/Guides/CollectionService

**Comment by YuuwaExpanderBleed** [comment]
I have the code ready, but I won't pr yet since I am not sure if this should be added or not.

**Comment by LastTalon** [comment]
I understand what's useful about this, but I want to make sure that hooks that make it into this library don't encourage non-idiomatic ECS usage.

My concern here is that if all you're intending to do is create components when they're tagged, the way the matter docs show should be the preferred way. I'm open to the possibility that this hook can be used for something other than that, but at the moment that's unclear to me.

These are the two things I'd like to see before adding something like this:
1. There are clear use-cases for this that aren't simply inserting a component that would represent the tag.
2. The specific interface of this hook won't discourage the usage of the method the matter docs provide.

**Comment by YuuwaExpanderBleed** [comment]
Hi, thank you for responding. 

I agree that at the moment, I am not so sure how else I would use this hook as I have never encountered such a case. However, I believe that being able to go through all tagged instances along with newly added instance once and only once on its addition or removal can be beneficial because it allows for matter system to also implement the following structure, which is useful since if game states are managed with the ecs loop, it makes more sense to do anything that is related to states from within the context too. 

```lua
-- Normally, outside of matter
local function applyChange(instance)
    -- do something with instance
end

for _, instance in CollectionService:GetTagged("someTag") do
    applyChange(instance)
end
CollectionService:GetInstanceAddedSignal("someTag"):Connect(applyChange)
CollectionService:GetInstanceRemovedSignal("someTag"):Connect(function() 
    -- handle removal
end)
```
```lua
-- What I propose
local function system(world, state)
    for _, collectionEvent: {added: boolean, removed: boolean, instance: Instance} in useCollection("someTag") do
        if collectionEvent.added then
            -- do something with instance, with state in mind if needed.
        elseif collectionEvent.removed then
            -- handle removal
        end
    end
end
```

I am not sure if this falls under the idiomatic ecs use because I am not too famliar with what the standard is, as matter is my first ever exposure to ecs. So if this is something that should be handled externally and not by hooks, I understand. 

As for number 2, I think this won't necessarily discourage the users from using the example in matter docs, because this is simply the common collection service pattern but with a different shape for it to be usable in matter system.

-- Edit
In hindsight, I admit that this may only serve niche uses because usually if we want to update instances with tags on events such as when state changes, we call getTagged for existing ones and InstanceAddedSignal for new ones. Meanwhile, this hook only exists to implement one time only change which might not really align with what ecs is.

**Comment by LastTalon** [comment]
> In hindsight, I admit that this may only serve niche uses because usually if we want to update instances with tags on events such as when state changes, we call getTagged for existing ones and InstanceAddedSignal for new ones. Meanwhile, this hook only exists to implement one time only change which might not really align with what ecs is.

I think this encapsulates the hesitation I have. In the case of `useStream` or `useEvent` you're using the hook to sync the ECS state and iteration with some external event as the events happen; you use it when you already have something in your ECS and are waiting for some signal from outside the ECS about how it should update at the correct time. With this, the primary use case I see is only the latter half of that, which is waiting for some signal from outside the ECS; if that's all we're doing we don't need to do that within our systems since we can access the world state from outside systems (in event driven code) just fine.

So for now I think we should avoid including something like this, but we can re-open this issue or a new issue later if a use case presents itself that requires that sort of synchronization of outside tagging and system iteration. The abstraction this hook is trying to create is valuable and can certainly be used alongside matter, I think it's just that making it a hook is the wrong choice, some other abstraction is likely better.
